docs/proposals/07-shoot-control-plane-migration.md

```diff
@@ -0,0 +1,236 @@
+The metadata is separated into two sections. The first is generated by the Gardener Controller Manager and then either used to generate new resources (e.g secrets) or is directly deployed to the Shoot's control plane on the __Destination Seed__.
+
+The second is generated by the extension controllers in the seed and then synced to the shoot by a new ShootState Synchronization Controller (described below).
+
+```
+apiVersion: core.gardener.cloud/v1alpha1
+kind: ShootState
+metadata:
+  name: my-shoot
+  namespace: garden-core
+gardener:
+  secrets:
+  - name: ca
+    data:
+      ca.crt: ...
+      ca.key: ...
+  - name: ssh-keypair
+    data:
+      id_rsa: ...
+  - name:
+...
+extensions:
+- kind: Infrastructure
+  state: ... (Terraform state)
+- kind: ControlPlane
+  purpose: normal
+  state: ... (Certificates generated by the extension)
+- kind: Worker
+  state: ... (Machine objects)
+```
+
+The state data is saved as a `runtime.RawExtension` type, which can be encoded/decoded by the corresponding extension controller.
+
+There can be sensitive data in the `ShootState` which has to be hidden from the end-users. Currently we plan to keep the `ShootState` resource encrypted in etcd. This means that the Gardener API server must have an encryption configuration for `ShootState` resources and the Gardener Controller Manager must have access to the encryption key.
+
+An alternative is to have an additional namespace for every project called `garden-<project-name>-state` in which to keep the `ShootState` resource.
+
+
+#### Size limitations
+
+There are limits on the size of the request bodies sent to the kubernetes API server when creating or updating resources: by default ETCD can only accept request bodies which do not exceed 1.5 MiB (this can be configured with the `--max-request-bytes` flag); the kubernetes API Server has a request body limit of 3 MiB which cannot be set from the outside (with a command line flag); the gRPC configuration used by the API server to talk to ETCD has a limit of 2 MiB per request body which cannot be configured from the outside; and `watch` requests have a 16 MiB limit on the buffer used to stream resources.
+
+This means that if `ShootState` is bigger than 1.5 MiB, the ETCD max request bytes will have to be increased. However, there is still an upper limit of 2 MiB imposed by the gRPC configuration.
+
+If `ShootState` exceeds this size limitation it must make use of configmap/secret references to store the state of extension controllers. This is an implementation detail of Gardener and can be done at a later time if necessary as extensions will not be affected.
+
+
+Spliting the `ShootState` into multiple resources could have a positive benefit on performance as the Gardener API Server and Gardener Controller Manager would handle multiple small resources instead of one big resource.
```

Comment:
```suggestion
Splitting the `ShootState` into multiple resources could have a positive benefit on performance as the Gardener API Server and Gardener Controller Manager would handle multiple small resources instead of one big resource.
```

Yes, considerations to take once we have some experience with the overall approach.